kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -271,48 +272,274 @@ void KvbAppFilter::readBlockRange(BlockId block_id_start,
+    LOG_ERROR(logger_, msg.str());
+    throw std::runtime_error(msg.str());
+  }
+  // update the offsets if we now have corresponding public/private event groups in storage
+  if (eg_data_state_->public_offset == 0) eg_data_state_->public_offset = public_start;
+  if (eg_data_state_->private_offset == 0) eg_data_state_->private_offset = private_start;
 
-  LOG_DEBUG(logger_, "readEventGroupRange event_group_id: " << event_group_id << " to " << event_group_id_end);
+  uint64_t event_group_id_end = private_end + public_end;
 
-  for (; event_group_id <= event_group_id_end; ++event_group_id) {
+  if (event_group_id_start == 0) {
+    throw InvalidEventGroupId(event_group_id_start);
+  }
+  if (event_group_id_start > event_group_id_end) {
+    throw InvalidEventGroupRange(event_group_id_start, event_group_id_end);
+  }
+
+  // populate and read global event group ids from eg_data_state_->event_group_id_batch in batches of size kBatchSize.
+  // For every global event group id received, lookup the data table to fetch the event group, filter and push the
+  // filtered even group to queue_out
+  while (eg_data_state_->curr_trid_event_group_id < event_group_id_end) {
+    LOG_DEBUG(logger_,
+              "Current tag_event_group_id: " << eg_data_state_->curr_trid_event_group_id
+                                             << ", event_group_id_end: " << event_group_id_end);
+    auto opt = getNextEventGroupId(client_id_, eg_data_state_);
+    uint64_t global_event_group_id;
+    if (opt.has_value()) {
+      global_event_group_id = opt.value();
+    } else {
+      std::stringstream msg;
+      msg << "No more event groups in storage";
+      LOG_WARN(logger_, msg.str());
+      break;
+    }
+    eg_data_state_->curr_trid_event_group_id++;
+    LOG_DEBUG(logger_,
+              "Global_event_group_id: " << global_event_group_id
+                                        << ", trid_event_group_id: " << eg_data_state_->curr_trid_event_group_id
+                                        << ", event_group_id_start: " << event_group_id_start);
+    // we are not at the starting point in the event group list yet, let's keep incrementing
+    // eg_data_state_->curr_trid_event_group_id until we reach the start
+    if (eg_data_state_->curr_trid_event_group_id < event_group_id_start) {
+      continue;
+    }
     std::string cid;
-    auto event_group = getEventGroup(event_group_id, cid);
+    auto event_group = getEventGroup(global_event_group_id, cid);
     if (event_group.events.empty()) {
       std::stringstream msg;
-      msg << "EventGroup doesn't exist for valid event_group_id: " << event_group_id;
+      msg << "EventGroup doesn't exist for valid event_group_id: " << global_event_group_id;
       throw KvbReadError(msg.str());
     }
-    KvbFilteredEventGroupUpdate update{event_group_id, filterEventsInEventGroup(event_group_id, event_group)};
+    KvbFilteredEventGroupUpdate update{eg_data_state_->curr_trid_event_group_id,
```

Comment:
From my understanding, we can never get an empty event group because we query public and private events and only get a reference to a global event group if one of the former points to it.
Therefore, can we assert this assumption here after filtering? `assert not update.event_group.events.empty()`
And similar in other places where we filter?

Reply:
Yes, thanks, I have added an assert in `filterEventsInEventGroup`.